lib/reader_writer.toit

```diff
@@ -80,27 +77,38 @@ monitor ReaderWriterHelper_:
     if writer_closed_: throw "CLOSED"
     result := to - from
     while from != to:
-      await: fullness_ != buffer_.size or outgoing_ == null or reader_closed_
+      await: fullness_ != buffer_size_ or reader_closed_
       if reader_closed_: throw "CLOSED"
+      // Allocate a buffer or reuse an old one.
+      if buffer_ == null:
+        if reuse_:
+          buffer_ = reuse_
+          reuse_ = null
+        else:
+          buffer_ = ByteArray buffer_size_
+
       space := buffer_.size - fullness_
-      if space == 0:
-        assert: outgoing_ == null
-        outgoing_ = buffer_
-        buffer_ = ByteArray buffer_size_
-        fullness_ = 0
-        space = buffer_.size
       remaining := to - from
       chunk_size := min space remaining
+      // Put as much as possible into the buffer.
+      // If we fill it up entirely, then the 'await' above will make us
+      // wait until a reader empties the buffer.
       buffer_.replace fullness_ data from from + chunk_size
       from += chunk_size
       fullness_ += chunk_size
     return result
 
   read:
-    await: outgoing_ or writer_closed_
-    if outgoing_:
-      result := outgoing_
-      outgoing_ = null
+    await: buffer_ or writer_closed_
+    result := ?
+    if buffer_:
+      if fullness_ != buffer_.size:
+        result = buffer_[..fullness_]
```

Comment:
I think this violates the expectation that read will return fresh byte arrays that can be held onto.  Subsequent calls to read will return arrays that alias each other.

Reply:
refactored.